// locizify/dist/commonjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _typeof2 = _interopRequireDefault(require("@babel/runtime/helpers/typeof"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _i18nextify = _interopRequireDefault(require("i18nextify"));

var _i18nextLocizeBackend = _interopRequireDefault(require("i18next-locize-backend"));

var _locizeEditor = _interopRequireDefault(require("locize-editor"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var i18next = _i18nextify["default"].i18next;
var enforce = {
  saveMissingTo: 'all'
};
var defaults = {
  reloadOnSave: true,
  bindSavedMissing: true
};
var reloadEditorOptions = {
  onEditorSaved: function onEditorSaved(lng, ns) {
    // location.reload();
    i18next.reloadResources(lng, ns, function () {
      _i18nextify["default"].forceRerender();
    });
  }
};
_i18nextify["default"].editor = _locizeEditor["default"];
i18next.use(_i18nextLocizeBackend["default"]).use(_locizeEditor["default"]);
var originalInit = i18next.init;

i18next.init = function () {
  var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var callback = arguments.length > 1 ? arguments[1] : undefined;
  options = _objectSpread(_objectSpread({}, defaults), options);
  var scriptEle = document.getElementById('locizify');

  if (scriptEle) {
    var config = {};
    var backend = {};
    var toRead = ['fallbackLng', 'saveMissing', 'debug', 'reloadOnSave', 'autorun', 'ele', 'cleanIndent', 'cleanWhitespace', 'namespace', 'namespaceFromPath'];
    var toReadAsArray = ['ignoreTags', 'ignoreIds', 'ignoreClasses', 'translateAttributes', 'mergeTags', 'inlineTags', 'ignoreInlineOn', 'ignoreCleanIndentFor', 'ns'];
    var toReadBackend = ['projectId', 'apiKey', 'referenceLng', 'version', 'allowedAddOrUpdateHost'];
    toRead.forEach(function (attr) {
      var value = scriptEle.getAttribute(attr.toLowerCase());
      if (value === 'true') value = true;
      if (value === 'false') value = false;
      if (value !== undefined && value !== null) config[attr] = value;
    });
    toReadAsArray.forEach(function (attr) {
      var value = scriptEle.getAttribute(attr.toLowerCase());
      if (value !== undefined && value !== null) config[attr] = value.split(',').map(function (item) {
        return item.trim();
      });
    });
    toReadBackend.forEach(function (attr) {
      var value = scriptEle.getAttribute(attr.toLowerCase());
      if (value === 'true') value = true;
      if (value === 'false') value = false;
      if (value !== undefined && value !== null) backend[attr] = value;
    });

    if (backend.allowedAddOrUpdateHost) {
      backend.allowedAddOrUpdateHosts = [backend.allowedAddOrUpdateHost];
      delete backend.allowedAddOrUpdateHost;
    }

    options = _objectSpread(_objectSpread(_objectSpread({}, defaults), options), config);
    options.backend = _objectSpread(_objectSpread({}, options.backend), backend);
  }

  if (options.reloadOnSave && (!options.editor || !options.editor.onEditorSaved)) options.editor = _objectSpread(_objectSpread({}, options.editor), reloadEditorOptions);

  if (options.bindSavedMissing) {
    options.backend.onSaved = function (lng, ns) {
      _locizeEditor["default"].handleSavedMissing(lng, ns);
    };
  }

  function handleI18nextInitialized(err, t) {
    // ready now
    // call orginal callback
    callback(err, t);
  }

  if (!options.backend.autoPilot || options.backend.autoPilot === 'false') return originalInit.call(i18next, _objectSpread(_objectSpread({}, options), enforce), handleI18nextInitialized);
  var locizeBackend = new _i18nextLocizeBackend["default"](options.backend);
  locizeBackend.getOptions(function (err, opts) {
    if (err && (typeof console === "undefined" ? "undefined" : (0, _typeof2["default"])(console)) === 'object' && typeof console.error === 'function') console.error(err);
    originalInit.call(i18next, _objectSpread(_objectSpread(_objectSpread({}, opts), options), enforce), handleI18nextInitialized);
  });
};

_i18nextify["default"].getLanguages = function (callback) {
  if (i18next.services.backendConnector) {
    i18next.services.backendConnector.backend.getLanguages(callback);
  } else {
    var ready = function ready() {
      i18next.off('initialized', ready);
      i18next.services.backendConnector.backend.getLanguages(callback);
    };

    i18next.on('initialized', ready);
  }
};

_i18nextify["default"].getOptions = function (callback) {
  if (i18next.services.backendConnector) {
    i18next.services.backendConnector.backend.getOptions(callback);
  } else {
    var ready = function ready() {
      i18next.off('initialized', ready);
      i18next.services.backendConnector.backend.getOptions(callback);
    };

    i18next.on('initialized', ready);
  }
};

var _default = _i18nextify["default"];
exports["default"] = _default;
