// i18nextify/dist/commonjs/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _i18next = _interopRequireDefault(require("i18next"));

var _i18nextHttpBackend = _interopRequireDefault(require("i18next-http-backend"));

var _i18nextBrowserLanguagedetector = _interopRequireDefault(require("i18next-browser-languagedetector"));

var _Observer = _interopRequireDefault(require("./Observer"));

var _docReady = _interopRequireDefault(require("./docReady"));

var _renderer = _interopRequireDefault(require("./renderer"));

var _missingHandler = require("./missingHandler");

var _utils = require("./utils");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function getDefaults() {
  return {
    autorun: true,
    ele: document.body,
    keyAttr: 'i18next-key',
    ignoreWithoutKey: false,
    ignoreTags: ['SCRIPT'],
    ignoreIds: [],
    ignoreClasses: [],
    translateAttributes: ['placeholder', 'title', 'alt', 'value#input.type=button', 'value#input.type=submit'],
    mergeTags: [],
    inlineTags: [],
    ignoreInlineOn: [],
    cleanIndent: true,
    ignoreCleanIndentFor: ['PRE', 'CODE'],
    cleanWhitespace: true,
    nsSeparator: '#||#',
    keySeparator: '#|#',
    debug: window.location.search && window.location.search.indexOf('debug=true') > -1,
    saveMissing: window.location.search && window.location.search.indexOf('saveMissing=true') > -1,
    namespace: false,
    namespaceFromPath: false,
    missingKeyHandler: _missingHandler.missingHandler,
    ns: [],
    onInitialTranslate: function onInitialTranslate() {}
  };
} // auto initialize on dom ready


var domReady = false;
var initialized = false;
(0, _docReady["default"])(function () {
  domReady = true;
  if (!initialized) init();
}); // extend i18next with default extensions

_i18next["default"].use(_i18nextHttpBackend["default"]);

_i18next["default"].use(_i18nextBrowserLanguagedetector["default"]); // log out missings
// i18next.on('missingKey', missingHandler);
// store last init options - for case init is called before dom ready


var lastOptions = {};

function changeNamespace(ns) {
  if (!ns && lastOptions.namespaceFromPath) ns = (0, _utils.getPathname)();
  lastOptions.ns.push(ns);
  lastOptions.defaultNS = ns;

  _i18next["default"].loadNamespaces(lastOptions.ns, function () {
    _i18next["default"].setDefaultNamespace(ns);
  });
}

var renderers = [];

function init() {
  var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  options = _objectSpread(_objectSpread(_objectSpread({}, getDefaults()), lastOptions), options);
  options = (0, _utils.parseOptions)(options); // delay init from domReady

  if (!options.ele) {
    delete options.ele;
    lastOptions = options;
  }

  initialized = true;
  var observer;

  function addRenderers(children) {
    for (var i = 0; i < children.length; i++) {
      var c = children[i];

      if (options.ignoreTags.indexOf(c.tagName) < 0 && options.ignoreIds.indexOf(c.id) < 0 && options.ignoreClasses.indexOf(c.className) < 0 && !c.attributes.localized && !c.attributes.translated) {
        var r = (0, _renderer["default"])(c, observer);
        renderers.push(r);
        r.render();
      }
    }
  }

  function waitForInitialRender(children, timeout, callback) {
    var allRendered = true;
    setTimeout(function () {
      for (var i = 0; i < children.length; i++) {
        var c = children[i];

        if (options.ignoreTags.indexOf(c.tagName) < 0 && options.ignoreIds.indexOf(c.id) < 0 && options.ignoreClasses.indexOf(c.className) < 0 && !c.attributes.localized && !c.attributes.translated) {
          if (allRendered) waitForInitialRender(children, 100, callback);
          allRendered = false;
          break;
        }
      }

      if (allRendered) callback();
    }, timeout);
  }

  var todo = 1;
  if (!domReady) todo++;
  if (options.autorun === false) todo++;

  function done() {
    todo -= 1;

    if (!todo) {
      if (!options.ele) options.ele = document.body;
      var children = options.ele.children;
      observer = new _Observer["default"](options.ele);
      addRenderers(children);
      observer.on('changed', function (mutations) {
        renderers.forEach(function (r) {
          return r.debouncedRender();
        });
        addRenderers(children);
      });
      waitForInitialRender(children, 0, function () {
        if (options.ele.style && options.ele.style.display === 'none') {
          options.ele.style.display = 'block';
        }

        options.onInitialTranslate();
      });
    }
  }

  _i18next["default"].init(options, done);

  if (!domReady) {
    (0, _docReady["default"])(done);
  }

  if (options.autorun === false) return {
    start: done
  };
}

function forceRerender() {
  renderers.forEach(function (r) {
    r.render(true); // enforce a rerender
  });
}

var _default = {
  init: init,
  i18next: _i18next["default"],
  changeNamespace: changeNamespace,
  forceRerender: forceRerender
};
exports["default"] = _default;
